tss-esapi/tests/integration_tests/abstraction_tests/transient_key_context_tests.rs

```diff
@@ -875,3 +889,78 @@ fn get_random_from_tkc() {
         .execute_without_session(|ctx| ctx.get_random(16))
         .expect("Failed to get random bytes");
 }
+
+#[test]
+fn sign_csr() {
+    // Check that we can convert a reference from TKC to Context
+    let mut ctx = create_ctx();
+
+    let (tpm_km, _tpm_auth) = ctx
+        .create_key(EcSigner::<NistP256>::key_params_default(), 0)
+        .expect("create private key");
+
+    let subject = Name::from_str("CN=tpm.example").expect("Parse common name");
+    let signer = EcSigner::<NistP256>::new(&mut ctx, tpm_km, None).expect("Create a signer");
+    let builder = RequestBuilder::new(subject, &signer).expect("Create certificate request");
+
+    let cert_req = builder
+        .build::<p256::ecdsa::DerSignature>()
+        .expect("Sign a CSR");
+
+    println!(
+        "{}",
+        cert_req
+            .to_pem(LineEnding::default())
+            .expect("Serialize CSR")
+    );
+}
+
+#[test]
+fn sign_p256_sha2_256() {
+    // Check that we can convert a reference from TKC to Context
+    let mut ctx = create_ctx();
+
+    let (tpm_km, _tpm_auth) = ctx
+        .create_key(EcSigner::<NistP256>::key_params::<sha2::Sha256>(), 0)
+        .expect("create private key");
+    let signer = EcSigner::<NistP256>::new(&mut ctx, tpm_km, None).expect("Create a signer");
+
+    let payload = b"Example of ECDSA with P-256";
+    let mut hash = Sha256::new();
+    hash.update(&payload);
+
+    let signature: p256::ecdsa::Signature = signer.sign_digest(hash.clone());
+    let verifying_key: VerifyingKey = signer.as_ref().clone();
+    assert!(verifying_key.verify_digest(hash, &signature).is_ok());
+}
+
+// NOTE(baloo): I believe this is a legitimate case, but for a reason I don't understand yet
+// the TPM will reply with `Error in creating derived key: 0x000002C3` when trying to create a
+// P-256 key with Sha3-256 associated hash.
+// This is combination defined by NIST:
+// https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/P256_SHA3-256.pdf
+//
+// I can't find where in the TPM spec this would be an illegal pair, we'll ignore for now.
+#[ignore]
+#[test]
+fn sign_p256_sha3_256() {
```

Comment:
This one is a bit puzzling to me. I can't find where in the spec the the P-256/SHA3-256 pair would be illegal. If anyone has an idea, I'd love to know.

Reply:
Could it be a limitation of the TPM you're using? From the [Structures spec](https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p59_Part2_Structures_pub.pdf), table 8, it seems SHA3-256 is classified as Assigned rather than TCG standard. I assume this might mean that manufacturers are not required to implement it (?)

Reply:
I'm using swtpm (0.8.2) with libtpms (0.9.6) I expected it to play nicely. I haven't took time to look in libtpms yet.

Reply:
Hmm, from [this](https://github.com/stefanberger/libtpms/blob/4b89a028dd55a73d0c36ba2a1d8c93a624994870/src/tpm2/TpmProfile_Common.h#L217) it seems libtpms can block SHA3 at compile-time, perhaps that's what is going on here.

Reply:
yeah, turns out it's more complicated than that.
https://github.com/stefanberger/libtpms/issues/206

No sha3 support for now.